pages/lua_manual/api_reference.md

```diff
@@ -0,0 +1,1849 @@
+### Methods
+
+| Return value | Name |
+| ------------ | ---- |
+| `Vector2f`{: .type }<br> | Rotate(`number`{: .type } ) |
+| `Vector2f`{: .type }<br> | Normalise() |
+| `number`{: .type }<br> | DotProduct(`Vector2f`{: .type } ) |
+| `Vector2f`{: .type} | new() |
+
+
+### Metafunctions
+
+| Metafunctions |
+| ------------- |
+| __add |
+| __sub |
+| __div |
+| __eq |
+| __mul |
+
+
+### Property Descriptions
+
+*  `number`{: .type } y
+
+> 
+
+*  `number`{: .type } magnitude
+
+> 
+
+*  `number`{: .type } x
+
+> 
+
+
+
+### Method Descriptions
+
+*  `Vector2f`{: .type } Rotate(`number`{: .type } )
+
+> 
+
+*  `Vector2f`{: .type } Normalise()
+
+> 
+
+*  `number`{: .type } DotProduct(`Vector2f`{: .type } )
+
+> 
+
+* `nil`{: .type} new()
+
+> 
+
+
+
+## ElementStyleProxy
+
+Inherits: `nil`{: .type }
+
+
+
+### Properties
+
+
+
+
+### Methods
+
+
+
+
+### Metafunctions
+
+| Metafunctions |
+| ------------- |
+| __newindex |
+| __ipairs |
+| __index |
+| __pairs |
+
+
+### Property Descriptions
+
+
+
+### Method Descriptions
+
+
+
+## ContextDocumentsProxy
+
+Inherits: `nil`{: .type }
+
+
+
+### Properties
+
+
+
+
+### Methods
+
+
+
+
+### Metafunctions
+
+| Metafunctions |
+| ------------- |
+| __pairs |
+| __index |
+| __ipairs |
+
+
+### Property Descriptions
+
+
+
+### Method Descriptions
+
+
+
+## SelectOptionsProxy
+
+Inherits: `nil`{: .type }
+
+
+
+### Properties
+
+
+
+
+### Methods
+
+
+
+
+### Metafunctions
+
+| Metafunctions |
+| ------------- |
+| __ipairs |
+| __index |
+| __pairs |
+
+
+### Property Descriptions
+
+
+
+### Method Descriptions
+
+
+
+## RmlUiContextsProxy
+
+Inherits: `nil`{: .type }
+
+
+
+### Properties
+
+
+
+
+### Methods
+
+
+
+
+### Metafunctions
+
+| Metafunctions |
+| ------------- |
+| __index |
+| __ipairs |
+| __pairs |
+
+
+### Property Descriptions
+
+
+
+### Method Descriptions
+
+
+
+## Log
+
+Inherits: `nil`{: .type }
+
+
+
+### Properties
+
+
+
+
+### Methods
+
+| Return value | Name |
+| ------------ | ---- |
+| `nil`{: .type } | LogMessage() |
+
+
+### Metafunctions
+
+
+
+
+### Property Descriptions
+
+
+
+### Method Descriptions
+
+* `nil`{: .type} Message()
+
+> 
+
+
+
+## ElementChildNodesProxy
+
+Inherits: `nil`{: .type }
+
+
+
+### Properties
+
+
+
+
+### Methods
+
+
+
+
+### Metafunctions
+
+| Metafunctions |
+| ------------- |
+| __ipairs |
+| __pairs |
+| __index |
+
+
+### Property Descriptions
+
+
+
+### Method Descriptions
+
+
+
+## Document
+
+Inherits: `Element`{: .type }
+
+Document derives from Element. Document has no constructor; it must be instantiated through a Context object instead, either by loading an external RML file or creating an empty document. It has the following methods and properties:
+
+### Properties
+
+| Types | Name |
+| ------------ | ---- |
+| `string`{: .type } | title |
+| `Context`{: .type } | context |
+
+
+### Methods
+
+| Return value | Name |
+| ------------ | ---- |
+| `nil`{: .type } | Hide() |
+| `nil`{: .type } | Show(`integer`{: .type } [flags]) |
```

Comment:
This now takes two enums as input.

Reply:
The binding code seems to define extra enums that are supposed to be sent into Document.Show() as arguments.

```cpp
template<> void ExtraInit<Document>(lua_State* L, int metatable_index)
{
    // some stuff

    //create the DocumentFocus table
    lua_getglobal(L,"DocumentModal");
    if(lua_isnoneornil(L,-1))
    {
        lua_pop(L,1); //pop unsucessful getglobal
        lua_newtable(L); //create a table for holding the enum
        lua_pushinteger(L,(int)ModalFlag::None);
        lua_setfield(L,-2,"None");
        lua_pushinteger(L,(int)ModalFlag::Modal);
        lua_setfield(L,-2,"Modal");
		lua_pushinteger(L, (int)ModalFlag::Keep);
        lua_setfield(L,-2,"Keep");
        lua_setglobal(L,"DocumentModal");
    }

	//create the DocumentFocus table
	lua_getglobal(L, "DocumentFocus");
	if (lua_isnoneornil(L, -1))
	{
		lua_pop(L, 1); //pop unsucessful getglobal
		lua_newtable(L); //create a table for holding the enum
		lua_pushinteger(L, (int)FocusFlag::None);
		lua_setfield(L, -2, "None");
		lua_pushinteger(L, (int)FocusFlag::Document);
		lua_setfield(L, -2, "Document");
		lua_pushinteger(L, (int)FocusFlag::Keep);
		lua_setfield(L, -2, "Keep");
		lua_pushinteger(L, (int)FocusFlag::Auto);
		lua_setfield(L, -2, "Auto");
		lua_setglobal(L, "DocumentFocus");
	}
}
```

And depending on how many arguments are passed into Show(), the correct Show() overload is called by the bindings:

```cpp
int DocumentShow(lua_State* L, Document* obj)
{
    int top = lua_gettop(L);
    if(top == 0)
        obj->Show();
    else if(top == 1)
    {
        ModalFlag modal = (ModalFlag)luaL_checkinteger(L,1);
        obj->Show(modal);
    }
	else
	{
        ModalFlag modal = (ModalFlag)luaL_checkinteger(L,1);
		FocusFlag focus = (FocusFlag)luaL_checkinteger(L,2);
		obj->Show(modal, focus);
	}
    return 0;
}
```

These sort of things might have been missed in the JSON

Reply:
I think there may be more problems like this in the API. I will try to fix all these once on my own by going through the bindings manually.

Reply:
Yeah, I think this needs to be fixed in manual review. 

Also, everything which has `lua_type` arguments means that you can pass in values of different types which will be resolved in the function (different function signatures if you will). You don't actually pass a type. These ones will probably need manual review as well.